File: src/extensions/java/GendersTest/GendersTest.java
/*****************************************************************************\
 *  Copyright (C) 2007-2019 Lawrence Livermore National Security, LLC.
 *  Copyright (C) 2001-2007 The Regents of the University of California.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Jim Garlick <garlick@llnl.gov> and Albert Chu <chu11@llnl.gov>.
 *  UCRL-CODE-2003-004.
 *
 *  This file is part of Genders, a cluster configuration database.
 *  For details, see <http://www.llnl.gov/linux/genders/>.
 *
 *  Genders is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *
 *  Genders is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *  details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with Genders.  If not, see <http://www.gnu.org/licenses/>.
\*****************************************************************************/

import gov.llnl.lc.chaos.*;

class GendersTest
{
    public static void main(String[] args)
    {
	try {

	    Genders gh = new Genders("GendersTest/testgenders");

	    try {
		System.out.println("NumNodes = " + gh.getnumnodes());
		System.out.println("Numattrs = " + gh.getnumattrs());
		System.out.println("Maxattrs = " + gh.getmaxattrs());
		System.out.println("Getnodname = " + gh.getnodename());
	    }
	    catch (Exception e) {
		System.out.println("unexpected exception: " + e);
	    }

	    String[] nodes;

	    try {
		nodes = gh.getnodes();
		for (String tmp : nodes)
		    System.out.println("GetNodes All Node: " + tmp);

		nodes = gh.getnodes("mgmt");
		for (String tmp : nodes)
		    System.out.println("GetNodes mgmt Node: " + tmp);

		nodes = gh.getnodes("foobarlalala");
		for (String tmp : nodes)
		    System.out.println("GetNodes foobarlalala Node: " + tmp);

		nodes = gh.getnodes("cfhost");
		for (String tmp : nodes)
		    System.out.println("GetNodes cfhost Node: " + tmp);

		nodes = gh.getnodes("cfhost", "foobar");
		for (String tmp : nodes)
		    System.out.println("GetNodes cfhost=foobar  Node: " + tmp);

		nodes = gh.getnodes("cfhost", "hypei");
		for (String tmp : nodes)
		    System.out.println("GetNodes cfhost=hypei  Node: " + tmp);
	    }
	    catch (Exception e) {
		System.out.println("unexpected exception: " + e);
	    }

	    String attrs[];
	    String vals[];

	    try {
		attrs = gh.getattr();
		for (String tmp : attrs)
		    System.out.println("getattr : " + tmp); 

		attrs = gh.getattr("hype355");
		for (String tmp : attrs)
		    System.out.println("getattr hype355: " + tmp); 

		attrs = gh.getattr_all();
		for (String tmp : attrs)
		    System.out.println("Getattr All : " + tmp);
	    }
	    catch (Exception e) {
		System.out.println("unexpected exception: " + e);
	    }

	    boolean b;

	    try {
		b = gh.isnode("foobar");
		System.out.println("Isnode foobar : " + b);

		b = gh.isnode("hype355");
		System.out.println("Isnode hype355 : " + b);

		b = gh.isnode(null);
		System.out.println("Isnode null : " + b);

		b = gh.isattr("foobar");
		System.out.println("Isattr foobar : " + b);

		b = gh.isattr("mgmt");
		System.out.println("Isattr mgmt : " + b);

		b = gh.isattrval("cfhost", "foobar");
		System.out.println("Isattrval cfhost=foobar : " + b);

		b = gh.isattrval("cfhost", "hypei");
		System.out.println("Isattrval cfhost=hypei : " + b);
	    }
	    catch (Exception e) {
		System.out.println("unexpected exception: " + e);
	    }

	    try {
		nodes = gh.query("mgmt");
		for (String tmp : nodes)
		    System.out.println("Query mgmt: " + tmp);

		nodes = gh.query("mgmt||login");
		for (String tmp : nodes)
		    System.out.println("Query mgmt||login: " + tmp);

		nodes = gh.query("mgmt&&login");
		for (String tmp : nodes)
		    System.out.println("Query mgmt&&login: " + tmp);

		nodes = gh.query("~mgmt");
		for (String tmp : nodes)
		    System.out.println("Query ~mgmt: " + tmp);

		b = gh.testquery("mgmt");
		System.out.println("TestQuery mgmt: " + b);

		b = gh.testquery("login");
		System.out.println("TestQuery login: " + b);

		b = gh.testquery("hype355", "mgmt");
		System.out.println("TestQuery mgmt: " + b);

		b = gh.testquery("hype355", "login");
		System.out.println("TestQuery login: " + b);

		b = gh.testquery("hype356", "mgmt");
		System.out.println("TestQuery mgmt: " + b);

		b = gh.testquery("hype356", "login");
		System.out.println("TestQuery login: " + b);
	    }
	    catch (Exception e) {
		System.out.println("unexpected exception: " + e);
	    }

	    String val;

	    try {
		val = gh.getattrval("mgmt");
		System.out.println("Getattrval mgmt : " + val);

		val = gh.getattrval("login");
		System.out.println("Getattrval login : " + val);

		val = gh.getattrval("cfhost");
		System.out.println("Getattrval cfhost : " + val);

		val = gh.getattrval("hype355", "mgmt");
		System.out.println("Getattrval hype355 mgmt : " + val);

		val = gh.getattrval("hype355", "login");
		System.out.println("Getattrval hype355 login : " + val);
	    }
	    catch (Exception e) {
		System.out.println("unexpected exception: " + e);
	    }

	    try {
		val = gh.getattrval("hype355", "cfhost");
		System.out.println("Getattrval hype355 cfhost : " + val);

		val = gh.getattrval("hype356", "mgmt");
		System.out.println("Getattrval hype356 mgmt : " + val);

		val = gh.getattrval("hype356", "login");
		System.out.println("Getattrval hype356 login : " + val);

		val = gh.getattrval("hype356", "cfhost");
		System.out.println("Getattrval hype356 cfhost : " + val);

		b = gh.testattr("mgmt");
		System.out.println("Testattr mgmt : " + b);

		b = gh.testattr("login");
		System.out.println("Testattr login : " + b);

		b = gh.testattr("cfhost");
		System.out.println("Testattr cfhost : " + b);

		b = gh.testattr("hype355", "mgmt");
		System.out.println("Testattr hype355 mgmt : " + b);

		b = gh.testattr("hype355", "login");
		System.out.println("Testattr hype355 login : " + b);

		b = gh.testattr("hype355", "cfhost");
		System.out.println("Testattr hype355 cfhost : " + b);

		b = gh.testattr("hype356", "mgmt");
		System.out.println("Testattr hype356 mgmt : " + b);

		b = gh.testattr("hype356", "login");
		System.out.println("Testattr hype356 login : " + b);

		b = gh.testattr("hype356", "cfhost");
		System.out.println("Testattr hype356 cfhost : " + b);

		b = gh.testattrval("mgmt", "");
		System.out.println("Testattrval mgmt= : " + b);

		b = gh.testattrval("mgmt", "foobar");
		System.out.println("Testattrval mgmt=foobar : " + b);

		b = gh.testattrval("login", "");
		System.out.println("Testattrval login= : " + b);

		b = gh.testattrval("login", "foobar");
		System.out.println("Testattrval login=foobar : " + b);

		b = gh.testattrval("cfhost", "hypei");
		System.out.println("Testattrval cfhost=hypei : " + b);

		b = gh.testattrval("cfhost", "foobar");
		System.out.println("Testattrval cfhost=foobar : " + b);

		b = gh.testattrval("hype355", "mgmt", "");
		System.out.println("Testattrval hype355 mgmt= : " + b);

		b = gh.testattrval("hype355", "mgmt", "foobar");
		System.out.println("Testattrval hype355 mgmt=foobar : " + b);

		b = gh.testattrval("hype355", "login", "");
		System.out.println("Testattrval hype355 login= : " + b);

		b = gh.testattrval("hype355", "login", "foobar");
		System.out.println("Testattrval hype355 login=foobar : " + b);

		b = gh.testattrval("hype355", "cfhost", "hypei");
		System.out.println("Testattrval hype355 cfhost=hypei : " + b);

		b = gh.testattrval("hype355", "cfhost", "foobar");
		System.out.println("Testattrval hype355 cfhost=foobar : " + b);

		b = gh.testattrval("hype356", "mgmt", "");
		System.out.println("Testattrval hype356 mgmt= : " + b);

		b = gh.testattrval("hype356", "mgmt", "foobar");
		System.out.println("Testattrval hype356 mgmt=foobar : " + b);

		b = gh.testattrval("hype356", "login", "");
		System.out.println("Testattrval hype356 login= : " + b);

		b = gh.testattrval("hype356", "login", "foobar");
		System.out.println("Testattrval hype356 login=foobar : " + b);

		b = gh.testattrval("hype356", "cfhost", "hypei");
		System.out.println("Testattrval hype356 cfhost=hypei : " + b);

		b = gh.testattrval("hype356", "cfhost", "foobar");
		System.out.println("Testattrval hype356 cfhost=foobar : " + b);
	    }
	    catch (Exception e) {
		System.out.println("unexpected exception: " + e);
	    }

	    try {
		gh.getattr("foobarnode");
	    }
	    catch (GendersExceptionNotfound e) {
		System.out.println("Got not found exception : " + e);
	    }
	    catch (Exception e) {
		System.out.println("unexpected exception: " + e);
	    }

	    gh.cleanup();

	    try  {
		System.out.println("NumNodes = " + gh.getnumnodes());
	    }
	    catch (GendersExceptionInternal e) {
		System.out.println("Got internal exception: " + e);
	    }
	    catch (Exception e) {
		System.out.println("unexpected exception: " + e);
	    }
	}
	catch (Exception e) {
	    System.out.println("unexpected exception: " + e);
	}
    }
}

File: src/extensions/lua/GendersTest/GendersTest.lua
#!/usr/bin/lua

db_file = "testgenders"

genders_lib = require("genders")
genders_handle = nil

-- wrapper to for using pcall
function wrapper_new() 
  genders_handle = genders_lib.new(db_file)
end

str_err = ""
str_out = ""

no_load = nil
error_status = 0

getter_funcs = {"getnumattrs","getnumnodes","getnodes"}
list_nodes = {"foobar","hype355","null"}
list_attr = {"foobar","mgmt"}
-- attr <-> value mismatch as this is dict!
list_attrval = {["foobaar"] = "cfhost", ["hypei"] = "cfhost"}
list_query = { "mgmt","mgmt||login","mhmt&&login","~mgmt"}

local open_lib, genders_handle = pcall(genders_lib.new,db_file,genders_lib)
if open_lib then
  print("Loaded database \""..db_file.."\"")
  for _, func_name in ipairs(getter_funcs) do 
    local str = "return function(handle) return pcall(handle."..func_name..",handle) end"
    local wrap_func = assert(load(str))
    local func = wrap_func()
    local ok,ret_val = func(genders_handle) 
    if ok then 
      if type(ret_val) ~= "table" then 
        print(func_name.." :: "..ret_val)
      else
        local tmp_str = func_name.." :: "
        for _,val in ipairs(ret_val) do
          tmp_str = tmp_str..val..", "
        end
        print(tmp_str)
      end
    else
      error_status = 1
      str_err = "Could not call "..func_name..": "..ret_val
    end
  end
  local ok, ret_val = pcall(genders_handle.getnodes,genders_handle,"mgmt")
  if ok then 
      local tmp_str = "getnodes(\"mgmt\") :: " for _,val in ipairs(ret_val) do tmp_str = tmp_str..val..", " end
      print(tmp_str)
  else
    error_status = 1
    str_err = "Could not call: getnodes(\"mgmt\")"..ret_val
  end
  local ok, ret_val = pcall(genders_handle.getnodes,genders_handle,"foobarlalala")
  if ok then 
      local tmp_str = "getnodes(\"foobarlalala\") :: " for _,val in ipairs(ret_val) do tmp_str = tmp_str..val..", " end
      print(tmp_str)
  else
    error_status = 1
    str_err = "Could not call: getnodes(\"foobarlalala\")"..ret_val
  end
  local ok, ret_val = pcall(genders_handle.getnodes,genders_handle,"cfhost","foobar")
  if ok then 
      local tmp_str = "getnodes(\"cfhost\",\"foobar\") :: " for _,val in ipairs(ret_val) do tmp_str = tmp_str..val..", " end
      print(tmp_str)
  else
    error_status = 1
    str_err = "Could not call: getnodes(\"cfhost\",\"foobar\")"..ret_val
  end
  local ok, ret_val = pcall(genders_handle.getnodes,genders_handle,"cfhost","hypei")
  if ok then 
      local tmp_str = "getnodes(\"cfhost\",\"hypei\") :: " for _,val in ipairs(ret_val) do tmp_str = tmp_str..val..", " end
      print(tmp_str)
  else
    error_status = 1
    str_err = "Could not call: getnodes(\"cfhost\",\"hypei\")"..ret_val
  end
  local ok, ret_val = pcall(genders_handle.getattr,genders_handle,"hype355")
  if ok then 
      local tmp_str = "getattr(\"hype355\") :: " for key,value in pairs(ret_val) do tmp_str = tmp_str..key if value == "" then tmp_str = tmp_str..", " else tmp_str = tmp_str.."="..value.."," end  end
      print(tmp_str)
  else
    error_status = 1
    str_err = "Could not call: getattr(\"hype355\")"..ret_val
  end
  for _,node in ipairs(list_nodes) do
    local ok, ret_val = pcall(genders_handle.isnode,genders_handle,node)
    if ok then
      if ret_val then print("isnode("..node..") = true") else print("isnode("..node..") = false") end
    else
      error_status = 1
      str_err = "Could not call: isnode("..node..") "..ret_val
    end
  end
  for _,attr in ipairs(list_attr) do
    local ok, ret_val = pcall(genders_handle.isattr,genders_handle,attr)
    if ok then
      if ret_val then print("isattr("..attr..") = true") else print("isattr("..attr..") = false") end
    else
      error_status = 1
      str_err = "Could not call: isattr("..attr..") "..ret_val
    end
  end
  for value,attr in pairs(list_attrval) do
    local ok, ret_val = pcall(genders_handle.isattrval,genders_handle,attr,value)
    if ok then
      if ret_val then print("isattrval("..attr..","..value..") = true") else print("isattrval("..attr..","..value..") = false") end
    else
      error_status = 1
      str_err = "Could not call: isattr("..attr..") "..ret_val
    end
  end
  for _,qry in ipairs(list_query) do
    local ok, ret_val = pcall(genders_handle.query,genders_handle,qry)
    if ok then
      local tmp_str = "query("..qry..") :: " for _,val in pairs(ret_val) do tmp_str = tmp_str..val..", " end
      print(tmp_str)
    else
      error_status = 1
      str_err = "Could not call: query("..qry..")"..ret_val
    end
  end
else
  str_err = "Failure in loading database \""..db_file.."\""
  error_status = 1
end
if error_status == 0 then
  print("No errors")
else
  print("Collected errors are: "..str_err)
end
